### PR TITLE
[Products] Product endpoints by code

### PR DESCRIPTION
**Pull request by lchrusciel**
Based on #178

Recreated PR #179, but from fork.

### CODE REVIEW

File: src/ViewRepository/ProductCatalogViewRepository.php [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+    /**
+     * @param string $localeCode
+     * @param PaginatorDetails $paginatorDetails
+     * @param ChannelInterface $channel
+     * @param TaxonInterface $taxon
+     *
+     * @return PageView
+     */
+    private function findByTaxon(string $localeCode, PaginatorDetails $paginatorDetails, ChannelInterface $channel, TaxonInterface $taxon): PageView
```
Comment (pamil) [inline src/ViewRepository/ProductCatalogViewRepository.php]: Shouldn't we preserve the same order as in `findByTaxon{Code,Slug}` methods? Taxon, locale, channel, paginator details?

File: src/ViewRepository/ProductDetailsViewRepository.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        ChannelRepositoryInterface $channelRepository,
+        ProductRepositoryInterface $productRepository,
+        ProductViewFactoryInterface $productViewFactory
+    ) {
+        $this->channelRepository = $channelRepository;
+        $this->productRepository = $productRepository;
+        $this->productViewFactory = $productViewFactory;
+    }
+
+    public function findOneBySlug(string $channelCode, string $productSlug, ?string $localeCode): ProductView
```
Comment (pamil) [inline src/ViewRepository/ProductDetailsViewRepository.php]: What about using some convention for ordering those parameters? It's always something related to subject, locale code and channel (code).

File: tests/Responses/Expected/product/product_review_list_page_by_slug_response.json [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            "author": "oliver@queen.com"
+        },
+        {
+            "title": "Nice mug",
+            "rating": @integer@,
+            "comment": "@string@",
+            "author": "oliver@queen.com"
+        },
+        {
+            "title": "Nice mug",
```
Comment (pamil) [inline tests/Responses/Expected/product/product_review_list_page_by_slug_response.json]: Lol, that mug's really nice. Or the reviews are counterfeited 🌮

File: src/Model/PaginatorDetails.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return $this->limit;
+    }
+
+    public function page(): int
+    {
+        return $this->page;
+    }
+
+    public function getParameters(): array
```
Comment (pamil) [inline src/Model/PaginatorDetails.php]: Something's wrong here:
 - if we don't use getters here, `addParameter` should be named `addGetParameter`
 - if we use getters here, `page` and `limit` methods should get the prefix

File: src/Model/PaginatorDetails.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    {
+        return $this->getParameters;
+    }
+
+    public function addParameter(string $key, string $value)
```
Comment (pamil) [inline src/Model/PaginatorDetails.php]: `: void`

Comment (pamil) [inline src/Model/PaginatorDetails.php]: Btw. `$value` can be an array as well.

File: doc/swagger.yml [hunk 6]
```diff
@@ -464,12 +464,60 @@ paths:
...
+              required: false
+              type: "integer"
+              example: 2
+            responses:
+                200:
+                    description: "Paginated product list."
+                    schema:
+                        type: "object"
+                        $ref: "#/definitions/ProductsPage"
+    /products-by-slug/{slug}:
```
Comment (lchrusciel) [inline doc/swagger.yml]: `/products/{code}` is typical rest format. We are not really RESTful, but on the other hand, everybody used to it.

File: src/Request/AddProductReviewByCodeRequest.php [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    /** @var string */
+    private $comment;
+
+    /** @var string */
+    private $email;
+
+    public function __construct(Request $request)
+    {
+        $this->slug = $request->attributes->get('code');
+
```
Comment (lchrusciel) [inline src/Request/AddProductReviewByCodeRequest.php]: slug is fetched from attributes, rest from the request. I just split them to improve readability.

File: src/Controller/Product/ShowProductDetailsByCodeAction.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        }
+
+        try {
+            return $this->viewHandler->handle(View::create($this->productCatalog->findOneByCode(
+                $request->query->get('channel'),
+                $request->attributes->get('code'),
+                $request->query->get('locale')
+            ), Response::HTTP_OK));
+        } catch (\InvalidArgumentException $exception) {
+            throw new NotFoundHttpException($exception->getMessage());
```
Comment (lchrusciel) [inline src/Controller/Product/ShowProductDetailsByCodeAction.php]: I just use Assert library to find proper result. We can open an issue with that to different between regular `InvalidArgumentException` and exceptions which could result in `NotFoundHttpException` in api context.

**Pull request merged**